std/digest/package.d

```diff
@@ -1212,3 +1212,141 @@ if (isInputRange!R1 && isInputRange!R2 && !isInfinite!R1 && !isInfinite!R2 &&
         assert(!secureEqual(hex1, hex2));
     }
 }
+
+/**
+ * Used to convert hex string to byte array
+ * Example:
+ * ---
+ * ubyte[2] sba = "0xAA".toDigest.staticArray!2;
+ * ubyte[] dby  = "0xBA".toDigest.array;
+ * ---
+ */
+auto toDigest(const string hex) pure nothrow @safe @nogc
+{
+    return HexToBytesConverter(hex);
+}
+
+///
+@safe unittest
+{
+    void testToDigest(const ubyte[] b, const string s) @safe
+    {
+        import std.conv : text;
+        import std.array : array;
+
+        assert(b == s.toDigest.array, text(s ~ ".toDigest returned ",
+                s.toDigest, ", instead of ", b));
+    }
+
+    testToDigest([255], "0xff");
+    testToDigest([], "");
+    testToDigest([], "0x");
+    testToDigest([0x01], "0x1");
+    testToDigest([255], "0xFf");
+    testToDigest([255], "ff");
+    testToDigest([0x0f, 0xff], "0xfff");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123AaAa");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123_AaAa");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123 AaAa");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123            AaAa");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123Aa______AA");
+    testToDigest([0x1, 0x23, 0xaa, 0xaa], "0x123AaGGAA");
+
+    void testToStaticDigest(size_t n)(const ubyte[n] b, const string s) @safe
+    {
+        import std.conv : text;
+        import std.array : staticArray;
+
+        ubyte[n] bb = s.toDigest.staticArray!n;
+        assert(b == bb, text(s ~ ".toDigest returned ", bb, ", instead of ", b));
+    }
+
+    testToStaticDigest!3([255, 0, 0], "0xff");
+    testToStaticDigest!3([0, 255, 0], "0x00ff");
+    testToStaticDigest!1([0x12], "0x12ff");
+    testToStaticDigest!2([0x12, 255], "0x12ff");
+    testToStaticDigest!2([0x12, 255], "0x12__ff");
+    testToStaticDigest!4([0x1, 0x23, 0xaa, 0xaa], "0x123AaGGAA");
+}
+
+private struct HexToBytesConverter
+{
+    import std.ascii : isHexDigit;
+
+    private string hex;
+    private size_t byteCount;
+    private size_t hexPos;
+    private size_t bytesPos;
+    private bool addZero;
+    private char[2] _front;
+
+    @safe @nogc pure nothrow
+    {
+        this(string t)
+        {
+            hex = t;
+            if (hex.length >= 2 && hex[0 .. 2] == "0x")
+            {
+                hex = hex[2 .. $];
+            }
+            foreach (c; hex)
+            {
+                byteCount += c.isHexDigit;
```

Comment:
Iterating through all the digits upfront is wasteful. You could do the `c.isHexDigit` in `front`/`popFront`.

Reply:
Also, I don't think the `byteCount` field is required. You could always check `c.length`

Reply:
> Iterating through all the digits upfront is wasteful. You could do the `c.isHexDigit` in `front`/`popFront`.

how to proceed with such strings like "0x1aa"
so if we don't count all bytes we could add bytes only to end like [0x1a, 0xa0];

but IMO when we see such a string we wanna parse it like [0x01, 0xaa]

Reply:
> Also, I don't think the `byteCount` field is required. You could always check `c.length`

I wanna filter non-hex values like '_'
f.e. to parse "0xffaa_ffaa"